### DISCUSSION TITLE
[Design Pattern] - 싱글톤 패턴

### DISCUSSION THREAD

**Discussion by Irisation23** [cs-collections/cs-collections-discussion#137] [m.Design-Pattern]
# [디자인패턴] 싱글톤 - 발표자료 Fin

# 0. TL;DR 📚

- 싱글톤 패턴을 도입한 클래스는 `단일 인스턴스`이다.
- `단일 인스턴스` 가 모두 공용으로 사용한다.
- `개방 폐쇄 원칙` 을 어길수 있다.
- `단일 책임 원칙` 을 어길수 있다.
- `멀티 스레드` 상황을 고려해야 한다.

---

# 1. 싱글톤이란? 🤔

- 싱글톤은 객체(인스턴스)를 한 개만 생성하는 패턴이다.

```jsx
‘한 개의 인스턴스만 가진다’
```

아래와 같이 해석될 수 있다. 

```jsx
‘인스턴스가 하나만 존재 해야한다.’
```

잠깐 객체지향으로 돌아와 우리가 배운 개념으로는 클래스는 `설계도`이다.

그 `설계도`로 만들어 내는 `여러 객체`를 통해 또 다른 설계도에서 만들어진 객체들과 프로그램의 세계에서 `메시지`를 주고 받으며 `상호작용` 하기를 기대한다.

객체가 하나면 우리가 알고있는 객체지향이 맞을까? 왜 싱글톤은 단일 객체를 가질 수 밖에 없었을까?

싱글톤에 대해 알아보자.

## 1.1 싱글톤 특징

- **메모리 절약**
    
    모든 객체는 JVM의 Heap 메모리에 저장된다.
    
    그러나 싱글턴 객체는 단 한번만 생성되기 때문에 최초에 생성된 단 한개의 객체만 Heap에 저장된다.
    
    이후 모든 클라이언트가 참조하는 해당 인스턴스는 Heap에 하나밖에 없는 주소를 참조한다.
    
- **데이터 공유**
    
    싱글톤으로 구현된 객체는 JVM Heap 에 저장된 하나의 객체를 모든 클라이언트들이 공유하게 된다.
    
    만약 싱글톤 객체에 공유할 정보를 넣는다면 손쉽게 모든 클라이언트들이 공유할 수 있게 된다.
    
- **멀티스레드 환경을 고려**
    
    다수의 클라이언트에서 하나의 객체, 즉 하나의 자원을 공유하는 상황이다 보니 자칫하면 치명적인 문제가 발생할 수 있다.  
    
    싱글톤 구현 시 멀티스레드를 고려하지 않으면 여러 클라이언트에서 하나의 자원에 접근하며
    
    **[[레이스 컨디션](https://ko.wikipedia.org/wiki/%EA%B2%BD%EC%9F%81_%EC%83%81%ED%83%9C)](https://ko.wikipedia.org/wiki/%EA%B2%BD%EC%9F%81_%EC%83%81%ED%83%9C)이 발생** 할 수 있다.
    
    만약 싱글턴의 생성 로직에서 레이스 컨디션이 발생한다면, 여러개의 객체가 생성되는 일도 발생할 수 있다.
    
- **테스트 코드 작성이 까다롭다.**
    
    테스트 케이스들을 작성할 때는 필연적으로 다양한 입력값, 출력값을 정의해야 한다.
    
    만약 싱글톤 클래스가 불변객체이고, 생성자를 통해서 값을 주입받는 다면?
    
    결론적으로 까다롭다라는 표현한 이유는 단순 테스트하기 위한 불필요한 코드가 증가하게 된다.
    

싱글톤은 이러한 특징들을 가지고 있다.

---

# 2. 싱글턴의 구현 방법

1. `public static final` 필드
2. 정적 팩터리
3. 열거 타입 활용(enum)
4. LazyHolder 방식
5. `volatile` 키워드

## 2.1 public static final 필드

```java
/**
 * 1. public static final 필드 방식
 */
public class Singleton1 {
    public static final Singleton1 instance = new Singleton1();
    private Singleton1() { }
}
```

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/24821c5c-361f-451f-94a5-257e5d5cb9b2/Untitled.png)

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/d0335f6f-7016-4033-891c-0ffde934e021/Untitled.png)

- static field는 Class Loader의 Initializing 단계에서 실제로 객체가 생성된다.
- Class Loader의 초기화 단계는 JVM 내부적으로 thread-safe 하게 동작하기 때문에 멀티스레드 환경에서도 안전하다.
- 하지만 이 싱글톤을 사용하지 않는다면 불필요한 메모리 사용이 일어난다. 이유는 static 변수인 `instance` GC의 대상이 아니기 때문이다.

## 2.2 **정적 팩터리 방식**

정적 팩터리 방식에서는 여러가지 방법을 사용할 수 있다.

이 방식들의 공통적인 특징은 다음과 같다.

- 통상적으로 `getInstance` 메소드는 싱글톤을 반환하는 기능을 한다고 알려져 있음
    - 그래서 클라이언트 측에서 싱글톤을 사용한다고 인지하고 사용할 수 있어서 혼동 가능성이 적음
- 그리고 상황에 따라 `getInstance` 메소드만 수정하면 `Singleton` <--> `Non Singleton` 으로 변경할 수 있음

### 1)  private static final 필드 사용

```java
/**
 * 2. 정적 팩터리 방식
 *     1) static final 필드 활용
 */
public class Singleton2_1 {
    private static final Singleton2_1 instance = new Singleton2_1();
    private Singleton2_1() { }

    public static Singleton2_1 getInstance() {
        return instance;
    }
}
```

2.1 과 특징이 같다.

### 2) Lazy Initialization(지연 초기화)

```java
/**
 * 2. 정적 팩터리 방식
 *     2) Lazy Initialization(지연 초기화)
 */
public class Singleton2_2 {
    private static Singleton2_2 instance;
    private Singleton2_2() { }

    public static Singleton2_2 getInstance() {
        if (instance == null) {
            instance = new Singleton2_2();
        }

        return instance;
    }
}
```

1)번 방법은 사용하지 않아도 불필요하게 객체가 생성된다는 단점이 있었음.

하지만 지연 초기화 방법을 사용하면 그러한 점을 해결할 수 있다.

- 호출할 때 null 인 경우에 객체가 생성되므로 불필요한 메모리 사용을 줄일 수 있음!
- 멀티스레드 환경에서 `레이스 컨디션`이 발생할 경우, 싱글톤을 만족하지 않을 수 있음!

### 3) Lazy Initialization + Synchronization(지연 초기화 + 동기화)

```java
/**
 * 2. 정적 팩터리 방식
 *     3) Lazy Initialization + Synchronization(지연 초기화 + 동기화)
 */
public class Singleton2_3 {
    private static Singleton2_3 instance;
    private Singleton2_3() { }

    public static synchronized Singleton2_3 getInstance() {
        if (instance == null) {
            instance = new Singleton2_3();
        }

        return instance;
    }
}
```

멀티스레드 환경에서 레이스 컨디션이 발생하는 경우를 방지하기 위해 동기화를 사용하는 방법임.

Java 의 synchronized 키워드를 이용해서 동기화 문제를 해결함.

- thread-safe 한 싱글톤임
- synchronized 키워드로 인해 성능이 저하될 수 있음

### 4) **LazyHolder 방식 - Initialization on Demand Holder (IODH) - [추천] ✨**

```java
/**
 * 4. LazyHolder 방식
 */
public class Singleton4 {

    private Singleton4() { }

    private static class LazyHolder {
        private static final Singleton4 INSTANCE = new Singleton4();
    }

    public static Singleton4 getInstance() {
        return LazyHolder.INSTANCE;
    }
}
```

LazyHolder 방식은 지연 초기화 방식의 성능적 문제를 개선하기 위해 제안된 방법

synchronized 키워드를 사용하면 JVM 내부적으로 락을 획득하는 작업 때문에 성능 저하가 발생

하지만 이 방법은 Class Loader 에 Singleton4 인스턴스 생성 역할을 맡기며 지연 초기화를 구현하는 방법임

클래스 로드와 초기화 과정은 thread-safe 하게 동작하는데, 이는 synchronized 보다는 성능이 보장됨.

- 지연 초기화로 인해 미사용 객체를 생성하지 않아 메모리를 효율적으로 사용
- synchronized 키워드를 사용한 방식보다는 성능상 이점이 있음

### 5) volatile + double-checked locking

```java
public class Singleton5 {
    private static volatile Singleton5 instance;

    private Singleton5() {}

    public static Singleton5 getInstance() {
        if (instance == null) { //1 
            synchronized (Singleton5.class) { //2 instance null 일때 동기화 블록 진입
                if (instance == null) {
                    instance = new Singleton5();
                }
            }
        }
        return instance;
    }
}
```

이 방식은 getInstance() 메서드가 호출될 때 인스턴스가 null인지 확인하고, null일 경우에만 동기화 블록에 진입한다. 

동기화 블록 내에서 다시 한 번 인스턴스가 null인지 확인하고, null일 경우에만 인스턴스를 생성한다.

- 해당 방식은 인스턴스가 이미 생성된 경우에는 동기화 블록에 진입하지 않기 때문에 성능 저하를 최소화할 수 있음.

`volatile` 키워드는 instance 변수가 여러 스레드에 의해 공유될 수 있음을 나타낸다. 

`volatile` 키워드는 변수의 값을 메모리에서 읽어오도록 강제하여, 여러 스레드가 동시에 접근할 경우 발생할 수 있는 문제를 방지한다.

# 3. `개방 폐쇄 원칙` 🔥 & `단일 책임 원칙` 🔥

자, 다시 1. 의 객체지향관점으로 돌아와서 생각 해 보자.

해당 `싱글턴 패턴` 은 `개방 폐쇄 원칙`와 `단일 책임 원칙`를 시원하게 어기고 있다. 객체지향의 5원칙 중 무려 2개를 어기고 있다.

- "소프트웨어 개체(클래스, 모듈, 함수 등)는 확장에 대해서는 열려 있어야 하지만 변경에 대해서는 닫혀 있어야 한다” - OCP
- 싱글톤 인스턴스가 너무 많은 일을 하거나 많은 데이터를 공유시킬 경우에 다른 클래스의 인스턴스들 간에 결합도가 높아져 "개방 폐쇄 원칙"을 위배하게 됨. - 변경에 닫혀있지 못함 너무 많은 연관관계 때문에

따라서 객체 지향 설계 원칙에 어긋나기 때문에 수정이 어려워지고 유지보수의 비용이 높아질 수 있다.

- 단일 책임 원칙을 어기는 예제

```java
public class Singleton {
    private static Singleton instance;

    private List<String> users;
    private List<String> products;

    private Singleton() {
        users = new ArrayList<>();
        products = new ArrayList<>();
    }

    public static Singleton getInstance() {
        if (instance == null) {
            instance = new Singleton();
        }
        return instance;
    }

    public void addUser(String user) {
        users.add(user);
    }

    public void removeUser(String user) {
        users.remove(user);
    }

    public List<String> getUsers() {
        return users;
    }

    public void addProduct(String product) {
        products.add(product);
    }

    public void removeProduct(String product) {
        products.remove(product);
    }

    public List<String> getProducts() {
        return products;
    }
}
```

- "한 클래스는 하나의 책임만 가져야 한다” - SRP
- 싱글톤 인스턴스가 너무 많은 일을 하거나 많은 데이터를 공유시킬 경우에 객체 지향적인 설계 원칙 중 하나인 "단일 책임 원칙"을 위반할 수 있다.

# 4. 결론 💬

- 소프트웨어 설계에서 자주 발생하는 문제들을 해결하기 위한 효율적인 방법론이라는 말을 곱씹으며, 도입을 하기에 앞서 구조를 살펴봐야 한다.
- 싱글턴은 예민하다.
- `멀티스레드`를 고려해야하고, `단일 책임 원칙`, `개방 폐쇄 원칙`을 잘 지켰는지 체크를 해야한다. ✅

# 5. 회고 🧹

2023-04-04 화

- 싱글턴 패턴을 구현함에 있어 꽤 많은 연관관계가 있단 사실을 알게 되었다.
- 개발을 하면서 코드를 작성하는 것보다는 (키보드에 손을 올리기 전에) 해당 기술이 왜 사용되는지에 대한 생각을 먼저 하는 것이 좋다고 느낀다.

- 출처
    1. [[https://medium.com/@joongwon/multi-thread-환경에서의-올바른-singleton-578d9511fd42](https://medium.com/@joongwon/multi-thread-%ED%99%98%EA%B2%BD%EC%97%90%EC%84%9C%EC%9D%98-%EC%98%AC%EB%B0%94%EB%A5%B8-singleton-578d9511fd42)](https://medium.com/@joongwon/multi-thread-%ED%99%98%EA%B2%BD%EC%97%90%EC%84%9C%EC%9D%98-%EC%98%AC%EB%B0%94%EB%A5%B8-singleton-578d9511fd42)
    2. [[이펙티브 자바](http://www.yes24.com/Product/Goods/65551284)](http://www.yes24.com/Product/Goods/65551284) - **아이템 3. private 생성자나 열거 타입으로 싱글턴임을 보증하라**
    3. [https://nesoy.github.io/articles/2018-06/Java-volatile](https://nesoy.github.io/articles/2018-06/Java-volatile)
    4. [https://inpa.tistory.com/entry/JAVA-%E2%98%95-JVM-%EB%82%B4%EB%B6%80-%EA%B5%AC%EC%A1%B0-%EB%A9%94%EB%AA%A8%EB%A6%AC-%EC%98%81%EC%97%AD-%EC%8B%AC%ED%99%94%ED%8E%B8](https://inpa.tistory.com/entry/JAVA-%E2%98%95-JVM-%EB%82%B4%EB%B6%80-%EA%B5%AC%EC%A1%B0-%EB%A9%94%EB%AA%A8%EB%A6%AC-%EC%98%81%EC%97%AD-%EC%8B%AC%ED%99%94%ED%8E%B8)
    5. [https://nesoy.github.io/articles/2018-06/Java-volatile](https://nesoy.github.io/articles/2018-06/Java-volatile)

**Comment by JoisFe** [comment]
LazyHolder 방식을 추천하신다고 하셨는데
 혹시 LazyHolder 방식의 단점은 없나요?
 
 그리고 `volatile 키워드는 변수의 값을 메모리에서 읽어오도록 강제하여, 여러 스레드가 동시에 접근할 경우 발생할 수 있는 문제를 방지한다.`
라고 했는데 변수의 값을 메모리에서 읽어오도록 강제하면 왜 동시성 문제를 방지할 수 있는지도 궁금합니다.

글 잘 읽었습니다.!

**Reply by Irisation23** [reply]
먼저 LazyHolder 방식의 단점으로 지목하신 키워드가 `IODH` 방식을 의미하는 것 같아서 답변드리자면

`IODH` 방식은 생성이 실패하지 않도록 보장되는 경우에만 사용할 수 있습니다. 
즉 [대부분의 JVM 구현에서 Something의 생성이 실패하면 이후 동일한 클래스 로더에서 초기화를 시도할 때 NoClassDefFoundError가 발생합니다 ](https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom)

> 따라서, Something(싱글턴 인스턴스)의 생성이 실패하지 않도록 보장되어야 합니다.

---

두번째 질문의 답변을 하기에 앞서 아래의 이미지를 참조해 주세요.

<img width="396" alt="image" src="https://user-images.githubusercontent.com/53285909/231327662-cacb7099-f421-499d-a348-6de7722c3f33.png">

각 스레드는 메모리의 변수를 참조하기 이전에 스레드 본인의 캐시를 먼저 참조하게 됩니다.
다시 말해 다른 스레드가 해당 변수의 값을 변경하더라도 해당 스레드가 읽어오는 값은 변경되지 않을 수 있습니다.

하지만 변수가 `volatile`로 선언된 경우, 해당 변수의 값이 변경될 때마다 `모든 코어의 캐시가 갱신`되므로 각 스레드가 `항상 최신의 값`을 읽어올 수 있습니다. 

> 따라서, `volatile` 키워드는 멀티 스레드 환경에서 변수의 가시성을 보장합니다.

```
변수의 가시성(visibility)이란 멀티 스레드 환경에서 하나의 스레드가 변수의 값을 변경했을 때 
다른 스레드가 해당 변경 사항을 언제 볼 수 있는지를 나타내는 개념입니다.
```

**Reply by JoisFe** [reply]
감사합니다!
